crates/admin/src/cluster_controller/scheduler.rs

```diff
@@ -73,9 +75,126 @@ impl<T: PartitionProcessorPlacementHints> PartitionProcessorPlacementHints for &
+        current: PartitionConfiguration,
+        next: Option<PartitionConfiguration>,
+    ) {
+        if self.latest_epoch_metadata_version < epoch_metadata_version {
+            self.latest_epoch_metadata_version = epoch_metadata_version;
+            self.current = current;
+            self.next = next;
+        }
+    }
+
+    fn replicas(&self) -> impl Iterator<Item = &PlainNodeId> {
+        self.current.replica_set.iter().chain(
+            self.next
+                .as_ref()
+                .map(|config| itertools::Either::Left(config.replica_set.iter()))
+                .unwrap_or(itertools::Either::Right(iter::empty())),
+        )
+    }
+
+    fn contains_replica(&self, node_id: PlainNodeId) -> bool {
+        self.current.replica_set.contains(node_id)
+            || self
+                .next
+                .as_ref()
+                .is_some_and(|config| config.replica_set.contains(node_id))
+    }
+
+    fn generate_instructions(
+        &self,
+        partition_id: &PartitionId,
+        observed_cluster_state: &ObservedClusterState,
+        commands: &mut BTreeMap<PlainNodeId, Vec<ControlProcessor>>,
+    ) {
+        // todo: Avoid cloning of node_set if this becomes measurable
+        let mut observed_state = observed_cluster_state
+            .partitions
+            .get(partition_id)
+            .map(|state| state.partition_processors.clone())
+            .unwrap_or_default();
+
+        if let Some(leader) = &self.leader {
+            if !observed_state
+                .remove(leader)
+                .is_some_and(|state| state.run_mode == RunMode::Leader)
+            {
+                commands.entry(*leader).or_default().push(ControlProcessor {
+                    partition_id: *partition_id,
+                    command: ProcessorCommand::from(RunMode::Leader),
+                });
+            }
+        }
+
+        for node_id in self.replicas() {
+            let run_mode = if self.leader == Some(*node_id) {
+                RunMode::Leader
+            } else {
+                RunMode::Follower
+            };
+            if !observed_state
+                .remove(node_id)
+                .is_some_and(|state| state.run_mode == run_mode)
+            {
+                commands
+                    .entry(*node_id)
+                    .or_default()
+                    .push(ControlProcessor {
+                        partition_id: *partition_id,
+                        command: ProcessorCommand::from(run_mode),
+                    });
+            }
+        }
+
+        // all remaining entries in observed_state are not part of target, thus, stop them!
+        for node_id in observed_state.keys() {
+            commands
+                .entry(*node_id)
+                .or_default()
+                .push(ControlProcessor {
+                    partition_id: *partition_id,
+                    command: ProcessorCommand::Stop,
+                });
```

Comment:
Adding the PartitionConfiguration version to the `ControlProcessor` would allow us to filter out outdated commands at the PPM.

Reply:
Done.